### PR TITLE
feat: support start/stop fragment marker

### PR DESCRIPTION
**Pull request by wyfo**
See https://github.com/eclipse-zenoh/zenoh/pull/1597

### CODE REVIEW

**Comment by wyfo** [comment]
All the code has been changed since I've started working on it... I'm sad.

File: src/protocol/codec/transport.c [hunk 1]
```diff
@@ -222,8 +268,8 @@ z_result_t _z_init_decode(_z_t_msg_init_t *msg, _z_zbuf_t *zbf, uint8_t header)
         msg->_cookie = _z_slice_empty();
     }
 
-    if ((ret == _Z_RES_OK) && (_Z_HAS_FLAG(header, _Z_FLAG_T_Z) == true)) {
-        ret |= _z_msg_ext_skip_non_mandatories(zbf, 0x01);
+    if ((ret == _Z_RES_OK) && _Z_HAS_FLAG(header, _Z_FLAG_T_Z) == true) {
```
Comment (sashacmc) [inline src/protocol/codec/transport.c:271]:
```suggestion
    if ((ret == _Z_RES_OK) && _Z_HAS_FLAG(header, _Z_FLAG_T_Z)) {
```

File: src/protocol/definitions/transport.c [hunk 2]
```diff
@@ -131,13 +139,25 @@ _z_transport_message_t _z_t_msg_make_init_syn(z_whatami_t whatami, _z_id_t zid)
...
         (msg._body._init._req_id_res != _Z_DEFAULT_RESOLUTION_SIZE)) {
         _Z_SET_FLAG(msg._header, _Z_FLAG_T_INIT_S);
     }
 
+#if Z_FEATURE_FRAGMENTATION == 1
+    bool has_patch = msg._body._join._patch != _Z_NO_PATCH;
+#else
+    bool has_patch = false;
+#endif
+    if (has_patch == true) {
```
Comment (sashacmc) [inline src/protocol/definitions/transport.c:157]:
```suggestion
    if (has_patch) {
```

File: src/protocol/definitions/transport.c [hunk 3]
```diff
@@ -153,13 +173,25 @@ _z_transport_message_t _z_t_msg_make_init_ack(z_whatami_t whatami, _z_id_t zid,
...
 
+#if Z_FEATURE_FRAGMENTATION == 1
+    bool has_patch = msg._body._init._patch != _Z_NO_PATCH;
+#else
+    bool has_patch = false;
+#endif
+    if (has_patch == true) {
+        _Z_SET_FLAG(msg._header, _Z_FLAG_T_Z);
+    }
+
```
Comment (sashacmc) [inline src/protocol/definitions/transport.c:194]:
```suggestion
#if Z_FEATURE_FRAGMENTATION == 1
    if (msg._body._init._patch != _Z_NO_PATCH) {
        _Z_SET_FLAG(msg._header, _Z_FLAG_T_Z);
    }
#endif
```

**Review by sashacmc — CHANGES_REQUESTED** [review]

File: src/protocol/codec/transport.c [hunk 4]
```diff
@@ -64,9 +64,14 @@ z_result_t _z_join_encode(_z_wbuf_t *wbf, uint8_t header, const _z_t_msg_join_t
...
     _Z_RETURN_IF_ERR(_z_zsize_encode(wbf, msg->_next_sn._val._plain._best_effort));
+#if Z_FEATURE_FRAGMENTATION == 1
+    bool has_patch = msg->_patch != _Z_NO_PATCH;
+#else
+    bool has_patch = false;
+#endif
     if (msg->_next_sn._is_qos) {
-        if (_Z_HAS_FLAG(header, _Z_FLAG_T_Z)) {
-            _Z_RETURN_IF_ERR(_z_uint8_encode(wbf, _Z_MSG_EXT_ENC_ZBUF | _Z_MSG_EXT_FLAG_M | 1));
+        if (_Z_HAS_FLAG(header, _Z_FLAG_T_Z) == true) {
```
[the remark this answers is not in the document]
Comment (wyfo) [reply src/protocol/codec/transport.c:73]: I asked you about before:
> Also, why do we use == false in zenoh-pico?

[the remark this answers is not in the document]
Comment (wyfo) [reply src/protocol/codec/transport.c:73]: I asked you about before:
> Also, why do we use == false in zenoh-pico?
> Just for more strict typing I guess... nvm 

and you answered with a 👍

[the remark this answers is not in the document]
Comment (sashacmc) [reply src/protocol/codec/transport.c]: Possible I answered 👍 on nvm.

I don't know why some legacy zenoh-pico code is written like this, but it doesn't add any readability, rather the opposite, it looks strange and inefficient (although of course, modern compilers can optimize it)

File: src/protocol/definitions/transport.c [hunk 5]
```diff
@@ -131,13 +139,25 @@ _z_transport_message_t _z_t_msg_make_init_syn(z_whatami_t whatami, _z_id_t zid)
...
 
+#if Z_FEATURE_FRAGMENTATION == 1
+    bool has_patch = msg._body._join._patch != _Z_NO_PATCH;
+#else
+    bool has_patch = false;
+#endif
+    if (has_patch) {
+        _Z_SET_FLAG(msg._header, _Z_FLAG_T_Z);
+    }
+
```
Comment (sashacmc) [inline src/protocol/definitions/transport.c:160]:
```suggestion
#if Z_FEATURE_FRAGMENTATION == 1
    if (msg._body._join._patch != _Z_NO_PATCH) {
        _Z_SET_FLAG(msg._header, _Z_FLAG_T_Z);
    }
#endif
```

**Review by sashacmc — CHANGES_REQUESTED** [review]

File: src/protocol/codec/transport.c [hunk 6]
```diff
@@ -388,25 +434,53 @@ z_result_t _z_fragment_encode(_z_wbuf_t *wbf, uint8_t header, const _z_t_msg_fra
     z_result_t ret = _Z_RES_OK;
     _Z_DEBUG("Encoding _Z_TRANSPORT_FRAGMENT");
     _Z_RETURN_IF_ERR(_z_zsize_encode(wbf, msg->_sn))
-    if (_Z_HAS_FLAG(header, _Z_FLAG_T_Z)) {
-        ret = _Z_ERR_MESSAGE_SERIALIZATION_FAILED;
+    if (msg->first) {
+        if (_Z_HAS_FLAG(header, _Z_FLAG_T_Z) == true) {
```
Comment (jean-roland) [inline src/protocol/codec/transport.c:438]: Just a nit and I see there are some that precedes that PR, but all conditions should be booleans so the `== true / false` would be redundant and can be removed.

Comment (wyfo) [reply src/protocol/codec/transport.c:438]: See https://github.com/eclipse-zenoh/zenoh-pico/pull/813#discussion_r1868961287

Comment (Mallets) [reply src/protocol/codec/transport.c:439]: @sashacmc @jean-roland @wyfo this is about MISRA C rules: https://stackoverflow.com/questions/44131637/misra-c-2012-rule-10-1-boolean-operand-to-be-used-in-case-where-expression-is-of

Comment (sashacmc) [reply src/protocol/codec/transport.c:439]: AFAIK, MISRA-C rules about using non-boolean values as boolean expression:
e. g.
```
int i = 1;
if (i) {}
```
but this kind of expression is absolutely legal:
```
bool b = true;
if (b) {}
```

Comment (Mallets) [reply src/protocol/codec/transport.c:439]: Well, yes but I also believe it depends on the static analyser... I'm not saying we should use `== true`, I was just providing a bit more context :)

**Review by sashacmc — APPROVED** [review]

**Comment by sashacmc** [comment]
> All CI is green but CI status checks which is stuck. I'm going to manual merge the PR.

This is a known issue, @diogomatsubara is working on.

**Pull request merged**